text/differential-layer-format.md

```diff
@@ -0,0 +1,143 @@
+expected value of ~ 4 bytes (depending on how often a page is modified; more
+frequent / less noizy WAL record generation = lower overhead).
+
+
+
+### Page versions array
+The Delta layer uses an in-memory sorted array of <lsn, pageversion> pairs to
+represent the changes to a page. This is correct and allows for O(log N) version
+lookups (for N = number of total page versions). This is OK, but can be improved
+upon: as mentioned in "Storage and file formats; Structures", we always
+need a base page image to reconstruct a page. By limiting the array of page
+versions stored for each page in the Layer to Branches, we eliminate a
+significant amount of entries from the search space; improving this search speed
+by O(log fraction of non-image page-versions) and reducing memory usage.
+
+Because we generally only require the latest version of a page, we store the newest
+versions of the page in the front of the array, so that the access to this data
+requires few extra operations.
+
+Assuming that layer files will not grow past a certain size (2^48 - 1 bytes),
+we can can use an 48-bit integer to store the location of a Branch.
+When in memory, with the 8-byte LSN, we have 2 bytes left to store some more
+metadata of the branch in the BranchPtr, such as type info and other relevant
+information.
+
+
+The per-page data layout of the Differential layer will thus be:
+
+```text
+  Page data:
+    ... metadata
+    page_versions: sorted array of (Lsn -> BranchPtr) pairs, in descending order of LSN
+      Lsn -> BranchPtr:
+        Lsn: 8 bytes
+        BranchPtr: 8 bytes
+          offset_to_branch: 6 bytes
+          branch type specific info: 2 bytes
+          ...
+      Lsn -> BranchPtr -> ...
+      Lsn -> BranchPtr -> ...
+      Lsn -> BranchPtr -> ...
+      ...
+```
+
+Note that the definition of BranchPtr is partially dependent on what data we
+want to store in the BranchPtr, and what data we want to store locally in the
+Branch. Either way, the overhead of one branch in the page versions array is
+16 bytes.
+
+
+
+### Page index
+The Delta layer does not use an index to check if a page exists in the Layer, but
+instead uses the combined index for on (page, pageversion). This means that finding
+a page in a Delta layer is log(n) (for n = number of page versions in the Layer).
+This is very expensive and not really sustainable when we're looking forward:
+larger segments means larger amounts of page versions in a layer, which means
+larger lookup times.  
+In the Delta layer there is also no description on when a page was last modified,
+so any failed lookup in the Delta layer must search back to the previous layer,
+making such lookups in the worst case O(log N * M) for N pageversions per layer
+and M layers. This currently is not too much of an issue, because this M is
+limited to 1 due to how our layers work: we always have an image layer
+immediately before the delta layer, which limits the number of Delta layers we
+access to 1.
+
```

Comment:
It will be nt true if we commit PR https://github.com/zenithdb/zenith/pull/1133

Reply:
That would expose the `O(M * log(N)` issue, yes.

Reply:
Yes